File: app/src/test/java/dev/bee/kanjianki/ManualSyncCoordinatorTest.java
package dev.bee.kanjianki;

import dev.bee.kanjianki.sync.ManualSyncEngine;
import dev.bee.kanjianki.sync.SyncProgress;

import org.junit.Test;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

public final class ManualSyncCoordinatorTest {
    @Test
    public void successfulSyncRunsSuccessActionBeforeRenderingResult() throws Exception {
        ManualSyncEngine.SyncResult result = syncResult(true, false, "ok");
        List<String> events = new ArrayList<>();
        AtomicReference<ManualSyncEngine.SyncResult> rendered = new AtomicReference<>();

        ManualSyncCoordinator coordinator = new ManualSyncCoordinator(
                directExecutor(),
                Runnable::run,
                progress -> {
                    events.add("run");
                    progress.onSyncProgress(SyncProgress.atStage(SyncProgress.Stage.READING_NOTES));
                    return result;
                },
                () -> events.add("success"),
                syncResult -> {
                    events.add("render");
                    rendered.set(syncResult);
                }
        );

        AtomicBoolean progressCalled = new AtomicBoolean(false);
        coordinator.start(update -> progressCalled.set(true));

        assertEquals(Arrays.asList("run", "success", "render"), events);
        assertTrue(progressCalled.get());
        assertSame(result, rendered.get());
    }

    @Test
    public void failedSyncRendersWithoutSuccessAction() throws Exception {
        ManualSyncEngine.SyncResult result = syncResult(false, false, "failed");
        AtomicBoolean success = new AtomicBoolean(false);
        AtomicReference<ManualSyncEngine.SyncResult> rendered = new AtomicReference<>();

        ManualSyncCoordinator coordinator = new ManualSyncCoordinator(
                directExecutor(),
                Runnable::run,
                progress -> result,
                () -> success.set(true),
                rendered::set
        );

        coordinator.start(null);

        assertFalse(success.get());
        assertSame(result, rendered.get());
    }

    private static Executor directExecutor() {
        return Runnable::run;
    }

    private static ManualSyncEngine.SyncResult syncResult(boolean success, boolean skipped, String message) throws Exception {
        Constructor<ManualSyncEngine.SyncResult> constructor = ManualSyncEngine.SyncResult.class.getDeclaredConstructor(
                boolean.class,
                boolean.class,
                int.class,
                int.class,
                String.class,
                String.class
        );
        constructor.setAccessible(true);
        return constructor.newInstance(success, skipped, 0, 0, message, "");
    }
}

File: app/src/test/kotlin/dev/bee/kanjianki/ManualSyncCoordinatorTest.kt
package dev.bee.kanjianki

import dev.bee.kanjianki.sync.ManualSyncEngine
import dev.bee.kanjianki.sync.SyncProgress
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Test
import java.lang.reflect.Constructor
import java.util.ArrayList
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

class ManualSyncCoordinatorTest {
    @Test
    fun successfulSyncRunsSuccessActionBeforeRenderingResult() {
        val result = syncResult(true, false, "ok")
        val events = ArrayList<String>()
        val rendered = AtomicReference<ManualSyncEngine.SyncResult>()

        val coordinator = ManualSyncCoordinator(
            directExecutor(),
            Runnable::run,
            { progress ->
                events.add("run")
                progress.onSyncProgress(SyncProgress.atStage(SyncProgress.Stage.READING_NOTES))
                result
            },
            ManualSyncCoordinator.SuccessAction { events.add("success") },
            { syncResult ->
                events.add("render")
                rendered.set(syncResult)
            },
        )

        val progressCalled = AtomicBoolean(false)
        coordinator.start { progressCalled.set(true) }

        assertEquals(listOf("run", "success", "render"), events)
        assertTrue(progressCalled.get())
        assertSame(result, rendered.get())
    }

    @Test
    fun failedSyncRendersWithoutSuccessAction() {
        val result = syncResult(false, false, "failed")
        val success = AtomicBoolean(false)
        val rendered = AtomicReference<ManualSyncEngine.SyncResult>()

        val coordinator = ManualSyncCoordinator(
            directExecutor(),
            Runnable::run,
            { result },
            ManualSyncCoordinator.SuccessAction { success.set(true) },
            rendered::set,
        )

        coordinator.start(null)

        assertFalse(success.get())
        assertSame(result, rendered.get())
    }

    private fun directExecutor(): Executor = Executor(Runnable::run)

    private fun syncResult(success: Boolean, skipped: Boolean, message: String): ManualSyncEngine.SyncResult {
        val constructor: Constructor<ManualSyncEngine.SyncResult> =
            ManualSyncEngine.SyncResult::class.java.getDeclaredConstructor(
                Boolean::class.javaPrimitiveType,
                Boolean::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                String::class.java,
                String::class.java,
            )
        constructor.isAccessible = true
        return constructor.newInstance(success, skipped, 0, 0, message, "")
    }
}
